File: androidApp/src/main/java/com/example/kmmsandbox/android/MyViewModel.kt
package com.example.kmmsandbox.android

import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.kmmsandbox.GreetingService
import com.example.kmmsandbox.MyException
import com.example.kmmsandbox.TickerService
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.launch

class MyViewModel : ViewModel() {
    // flow

    private val tickerService = TickerService()
    private var _tickerFlow = MutableStateFlow<Int?>(null)
    private var tickerJob: Job? = null

    val tickerFow: StateFlow<Int?> = _tickerFlow
    private val _tickerErrorMessage = mutableStateOf("")
    val tickerErrorMessage: State<String> get() = _tickerErrorMessage

    fun startTicker() {
        _tickerErrorMessage.value = ""
        cancelTicker()
        tickerJob = viewModelScope.launch {
            tickerService.launchTickFlow()
                .catch { throwable: Throwable ->
                    if (throwable is MyException) {
                        _tickerErrorMessage.value = "'tickerFlow' error: ${throwable.localizedMessage}"
                    }
                }
                .collect {
                    _tickerFlow.value = it
                }
        }
    }

    fun cancelTicker() {
        tickerJob?.cancel()
        _tickerFlow.value = null
    }


    // suspend function

    private val greetingService = GreetingService()

    private var greetingJob: Job? = null

    val greetingText = mutableStateOf("")

    private val _greetingErrorMessage = mutableStateOf("")
    val greetingErrorMessage: State<String> get() = _greetingErrorMessage

    fun loadGreeting() {
        _greetingErrorMessage.value = ""
        cancelGreeting()
        greetingJob = viewModelScope.launch {
            greetingText.value = "Loading..."
            try {
                greetingText.value = greetingService.greet()
            } catch (e: MyException) {
                _greetingErrorMessage.value = "'greet' call error: ${e.localizedMessage}"
            }
        }
    }

    fun cancelGreeting() {
        if (greetingJob?.isCompleted != true) {
            greetingText.value = "Loading canceled"
        }
        greetingJob?.cancel()
    }

    val progressFlow: Flow<String?>
        get() = greetingService.progressFlow

    // error mode
    private val _errorMode = mutableStateOf(false)
    val errorMode: State<Boolean> get() = _errorMode
    fun changeErrorMode(value: Boolean) {
        _errorMode.value = value
        tickerService.changeErrorMode(value)
        greetingService.changeErrorMode(value)
    }
}

File: iosApp/iosApp/MyViewModel.swift
//
//  GreetingViewModel.swift
//  iosApp
//
//  Created by Svetlana.Isakova on 26.04.23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared
import KMPNativeCoroutinesCore
import KMPNativeCoroutinesAsync
import KMPNativeCoroutinesCombine

@MainActor
class MyViewModel: ObservableObject {
    
    // flow
    
    private let tickerService = TickerService()
    private var tickerTask: Task<(), Never>? = nil
    
    @Published
    var tickValue: Int? = nil
    
    @Published
    var tickerErrorMessage: String = ""
    
    func startTicker() {
        tickerErrorMessage = ""
        cancelTicker()
        tickerTask = Task {
            do {
                let sequence = asyncSequence(for: tickerService.launchTickFlow())
                for try await tick in sequence {
                    tickValue = tick.intValue
                }
            } catch {
                tickValue = nil
                tickerErrorMessage = "'launchTickFlow' error: \(error.localizedDescription)"
            }
        }
    }
    
    func cancelTicker() {
        tickerTask?.cancel()
        tickValue = nil
    }
    
    
    // suspend function
    
    private let greetingService = GreetingService()
    
    private var greetingTask: Task<(), Never>? = nil
    
    @Published
    var greetingText: String = ""
    
    @Published
    var greetingErrorMessage: String = ""
    
    func loadGreeting() {
        greetingErrorMessage = ""
        cancelGreeting()
        greetingTask = Task {
            greetingText = "Loading..."
            // alternative: to use do-catch and
            // try await asyncFunction(for: greetingService.greet())
            
            let result = await asyncResult(for: greetingService.greet())
            switch result {
            case let.success(text):
                greetingText = text
            case let.failure(error):
                if !(error is CancellationError)  {
                    greetingErrorMessage = "'greet' call error: \(error.localizedDescription)"
                }
            }
        }
    }

    func cancelGreeting() {
        greetingText = "Loading canceled"
        greetingTask?.cancel()
    }

    @Published
    var progress: String? = nil
    
    init() {
        Task {
            do {
                let sequence = asyncSequence(for: greetingService.progressFlow)
                for try await value in sequence {
                    progress = value
                }
            } catch {
                greetingErrorMessage = "'progressFlow' error: \(error.localizedDescription)"
                progress = nil
            }
        }
    }
    
    // error mode
    
    func changeErrorMode(value: Bool) {
        tickerService.changeErrorMode(value: value)
        greetingService.changeErrorMode(value: value)
    }
}
